flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/statistics/ContinuousHistogram.scala

```diff
@@ -0,0 +1,315 @@
+      return s.round.toInt
+    } else if (index == bins - 1) {
+      m_b = getCounter(index) +
+        (-getCounter(index)) * (b - getValue(index)) / (upper - getValue(index))
+      s = (getCounter(index) + m_b) *
+        (b - getValue(index)) / (2 * (upper - getValue(index)))
+    } else {
+      m_b = getCounter(index) + (getCounter(index + 1) - getCounter(index)) *
+        (b - getValue(index)) / (getValue(index + 1) - getValue(index))
+      s = (getCounter(index) + m_b) *
+        (b - getValue(index)) / (2 * (getValue(index + 1) - getValue(index)))
+    }
+    for (i <- 0 to index - 1) {
+      s = s + getCounter(i)
+    }
+    s = s + getCounter(index) / 2
+    s.round.toInt
+  }
+
+  /** Returns the total number of entries in the histogram
+    *
+    * @return Total number of points added in the histogram
+    */
+  override def total: Int = data.iterator.map(_._2).sum
+
+  /** Bin counter access function
+    *
+    * @param bin bin number to access
+    * @return `m_bin` = counter of bin
+    */
+  private[statistics] def getCounter(bin: Int): Int = {
+    require(0 <= bin && bin < bins, bin + " not in [0, " + bins + ")")
+    data.apply(bin)._2
+  }
+
+  /** Number of bins currently being used
+    *
+    * @return number of bins
+    */
+  override def bins: Int = {
+    data.size
+  }
+
+  /** Returns the string representation of the histogram.
+    *
+    */
+  override def toString: String = {
+    s"Size:" + bins + " " + data.toString
+  }
+
+  /** Loads values and counters into the histogram.
+    * This action can only be performed when there the histogram is empty
+    *
+    * @param counts Array of tuple of (value,count)
+    */
+  def loadData(counts: Array[(Double, Int)]): Unit = {
+    if (data.nonEmpty) {
+      throw new RuntimeException("Data can only be loaded during initialization")
+    }
+    data ++= counts
+    require(checkSanity, "Invalid initialization from data")
+  }
+
+  /** Bin value access function
+    *
+    * @param bin bin number to access
+    * @return `v_bin` = value of bin
+    */
+  private[statistics] def getValue(bin: Int): Double = {
+    require(0 <= bin && bin < bins, bin + " not in [0, " + bins + ")")
+    data.apply(bin)._1
+  }
+
+  /** Checks whether the values loaded into data by loadData are properly sorted or not.
+    * All values should be in (lower,upper)
+    * All counters loaded into data by loadData should be positive
+    *
+    */
+  private def checkSanity: Boolean = {
+    val size = bins
+    if (size == 0) return true
+    if (lower >= getValue(0)) return false
+    if (upper <= getValue(size - 1)) return false
+    for (i <- 0 to size - 2) {
+      if (getValue(i + 1) < getValue(i)) return false // equality will get merged later on
+      if (getCounter(i) <= 0) return false
+    }
+    if (getCounter(size - 1) <= 0) return false
+    // now bring the histogram to capacity and merge all <i>very-close</i> elements
+    while (mergeElements()) {}
```

Comment:
This is somehow hidden in the checkSanity method. Makes the behaviour of the method somewhat surprising.